### PR TITLE
warn on unused controls/parsers

### PR DESCRIPTION
**Pull request by grg**
Add warnings for unused P4Control and P4Parser objects to match P4Table warnings. Addresses issue #4439.

### CODE REVIEW

File: testdata/p4_16_samples_outputs/issue3379-1.p4-stderr
Comment (jafingerhut) [inline testdata/p4_16_samples_outputs/issue3379-1.p4-stderr:6]: @grg Do you know why Myparser1 is considered unused in this test program?

**Review by jafingerhut — APPROVED** [review]
I left one question about a warning message for one of the test programs, but if there is a good explanation for that one, the rest all looks good to me.

File: testdata/p4_16_samples_outputs/issue3379-1.p4-stderr
Comment (fruffy) [reply testdata/p4_16_samples_outputs/issue3379-1.p4-stderr:6]: I believe that is because there is no `main` declaration.

**Comment by grg** [comment]
@jafingerhut / @fruffy: Any thoughts on whether these warnings should be added? I opened a ticket (#4439) to get feedback on this. My opinion is that the extra warnings should be added for consistency.

**Comment by fruffy** [comment]
> @jafingerhut / @fruffy: Any thoughts on whether these warnings should be added? I opened a ticket (#4439) to get feedback on this. My opinion is that the extra warnings should be added for consistency.

I do not have any strong opinions but I think your change makes sense.

But what about multi-pipe programs? Do the warnings work correctly here? I guess they should since the parser/control is also removed.

**Comment by grg** [comment]
> > @jafingerhut / @fruffy: Any thoughts on whether these warnings should be added? I opened a ticket (#4439) to get feedback on this. My opinion is that the extra warnings should be added for consistency.
> 
> I do not have any strong opinions but I think your change makes sense.
> 
> But what about multi-pipe programs? Do the warnings work correctly here? I guess they should since the parser/control is also removed.

It works as expected with multi-pipe programs: a warning is emitted only if a control/parser is _never_ used in any of the pipes. If it's used in at least one pipe then no warning is emitted.

**Review by fruffy — APPROVED** [review]

**Comment by grg** [comment]
@asl / @ChrisDodd Is this proposed change likely to impact either of you negatively?

**Comment by asl** [comment]
> @asl / @ChrisDodd Is this proposed change likely to impact either of you negatively?

I believe we are ok. Cross-checking with @kfcripps

**Comment by jafingerhut** [comment]
I see 3 approvals on this.  Are we waiting for explicit approval from @ChrisDodd ?

**Comment by grg** [comment]
> I see 3 approvals on this. Are we waiting for explicit approval from @ChrisDodd ?

I've added it to the merge queue. This was put on the back burner last week while I was working on upstreaming the midend def-use pass from Tofino.

**Pull request merged**
